### PR TITLE
Adding testing to gdscript module

### PR DESCRIPTION
**Pull request by mashumafi**
Currently the need for a good testing solution is desired in the community. There are a few solutions such as [GUT](https://github.com/bitwes/Gut) and now [WAT](https://github.com/CodeDarigan/WAT). While GUT acted as a separate scene you can run with tests, it does not seem to integrate with the editor. WAT tried to address this with a editor focused testing but similar functionality of GUT. I am proposing another solution: that testing be native plugin that comes with GDScript. This is what this PR addresses. While not complete and probably not even what I expect to be the final solution I created a starting point for another test suite. Based off what python unittest module provides the goal is to provide a similar interface. It is part of the gdscript module and only in editor builds. Here is a summary of the classes added to support testing through this plugin. 

`TestConfig` is a singleton for accessing all configurable test options which can be changed in the project settings
`TestLoader` is a class to recursively search the project directory for tests and creates a `TestSuite`
`TestRunner` is a `MainLoop` that uses `TestLoader` to load and then runs the generated `TestSuite`
`MethodWatcher` is a provides and API that can be used to track or redirect what a method does along with some helper functions to check what those calls looks like alone with affecting their return values
`SignalWatcher` tracks which signals are called and their arguments and provides some helper checks
`TestCase` is meant to be inherited by test files and provides asserts and watcher for signals and mock
`ProxyScript` provide a script that proxys both `Script` and `ScriptInstance` of a class with a `MethodWatcher` to override and track which methods were called
`TestState` keeps track of the currently executing test along with other info, it's used to allow for yielding in test scripts by using `TestCase` similarly how `MainLoop` iterate method works.
`TestLog` keeps track of all failing asserts and logs that were written during the test suite
`TestResult` tracks all results from the running tests
`TextTestResult` same as `TestResult` but includes a log
`TestPlugin` provides the button right next to the other play button to run the tests, currently there a few hacks in place to create a tracked PID similar to how pressing play works. After tests run it also provides a panel to show the log results.

This is just a quick run-down of workflow. The goal is to at least point out that testing is a feature that is important to godot and having it in the engine give a lot more flexibility that a GDScript plugin cannot provide. For example, I can allow break points to hit on failed asserts because of the engine integration. Also I could imagine code coverage and documentation testing being included too. Thanks for taking a look and any questions related to this project or testing can be discussed further.

### CODE REVIEW

**Comment by vnen** [comment]
I don't think this should be integrated with the engine. Test suites tend to be opinionated and spread in different formats. I'd rather expose more hooks in the API for what's needed to make them as plugins.

**Comment by mashumafi** [comment]
Hey, good point and thanks for the response. Maybe I could help tackle adding hooks to help out with some of the issues I ran into. I will close this PR.

**Comment by CodeDarigan** [comment]
@vnen @mashumafi 

This may deserve its own issue but I've only just seen it here.

On the topic of exposing more hooks. One of the biggest stumbling blocks I've come across when creating WAT is dealing with Inner Classes.

The way I create Test Doubles is by collecting the source_code property of a script. 

```
var source: String = ""
var script = script
while script != null: // The null check is to stop at built-in classes
    source += script.source_code
    script = script.get_base_script()
// other
```

Inner Classes have a source_code property but it always returns empty. I'm not sure if this is a bug or just a consequence of being an inner class but it presents a problem. In the case above, if get_base_script() returns an inner class, then we've just missed out on some source_code;

This leads to odd issues where we have to instance it, then inst2dict it, to collect the sub paths. 

```
var data = inst2dict(script.new())
var path = data["@path"]
var subpath = data["@path"]
// We go to the main script
// Split by newlines
// Find line that contains our intended inner class
// Eliminate all lines that have a lower indent
// Collect all lines that the indent level of our inner_class + one more
```

1) Why do we have to instance something to get its sub path? Wouldn't it make more sense to expose it 
as a property? or a bool option to include sub_paths when getting the .resource_path

2) We can't do const script = preload("path.gd").inner (I'm guessing because inner is dynamic data). This is currently forcing me to have users pass in ("string.gd, "inner/class/list") to help prevent any unexpected nulls.

3) A huge reason for doing a lot of this is simply so we can get the inner classes of another class. I see in GDScript there is a subclasses map, so maybe if we could expose that as a Dict it would be a huge benefit.

Basically my point is
1) Expose Inner Classes
2) Give them actual source_code 

Then I'll be able to do a lot more with WAT.
